pgvectorscale/src/access_method/labels.rs

```diff
@@ -0,0 +1,271 @@
+        labels.sort_unstable();
+        labels.dedup();
+
+        Self { labels }
+    }
+}
+
+/// Accessor trait to allow common logic between `LabelSet` and `ArchivedLabelSet`.
+pub trait LabelSetView {
+    fn labels(&self) -> &[Label];
+
+    /// Any labels in the labelset?
+    fn is_empty(&self) -> bool {
+        self.labels().is_empty()
+    }
+
+    /// Do the labelsets share any labels?
+    fn overlaps<T: LabelSetView>(&self, other: &T) -> bool {
+        let a = self.labels();
+        let b = other.labels();
+
+        debug_assert!(a.is_sorted());
+        debug_assert!(b.is_sorted());
+
+        let mut i = 0;
+        let mut j = 0;
+
+        while i < a.len() && j < b.len() {
+            match a[i].cmp(&b[j]) {
+                std::cmp::Ordering::Equal => return true,
+                std::cmp::Ordering::Less => i += 1,
+                std::cmp::Ordering::Greater => j += 1,
+            }
+        }
+        false
+    }
+
+    fn iter(&self) -> std::slice::Iter<Label> {
+        self.labels().iter()
+    }
+}
+
+impl LabelSetView for LabelSet {
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+impl LabelSetView for ArchivedLabelSet {
+    fn labels(&self) -> &[Label] {
+        &self.labels
+    }
+}
+
+impl Debug for ArchivedLabelSet {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_tuple("ArchivedLabelSet")
+            .field(&self.labels.to_vec())
+            .finish()
+    }
+}
+
+/// A labeled vector is a vector with an optional set of labels.
+#[derive(Debug)]
+pub struct LabeledVector {
+    vec: PgVector,
+    labels: Option<LabelSet>,
+}
+
+impl LabeledVector {
+    pub fn new(vec: PgVector, labels: Option<LabelSet>) -> Self {
+        Self { vec, labels }
+    }
+
+    pub unsafe fn from_datums(
+        values: *mut Datum,
+        isnull: *mut bool,
+        meta_page: &MetaPage,
+    ) -> Option<Self> {
+        let vec = PgVector::from_pg_parts(values, isnull, 0, meta_page, true, false)?;
+
+        let labels: Option<LabelSet> = if meta_page.has_labels() {
+            if *isnull.add(1) {
+                Some(LabelSet::default())
+            } else {
+                let arr = Array::<i16>::from_datum(*values.add(1), false);
+                Some(arr.map_or_else(LabelSet::default, |arr| {
+                    // Special case to work around apparent bug in pgrx
+                    if arr.is_empty() || arr.iter().all(|x| x.is_none()) {
+                        return LabelSet::default();
+                    }
+                    let labels_iter = arr.into_iter().flatten();
+                    labels_iter.collect()
+                }))
+            }
+        } else {
+            None
+        };
+
+        Some(Self::new(vec, labels))
+    }
+
+    pub unsafe fn from_scan_key_data(
+        keys: &[ScanKeyData],
+        orderbys: &[ScanKeyData],
+        meta_page: &MetaPage,
+    ) -> Self {
+        let query = unsafe {
```

Comment:
what happens if the labels are in the order by. Do they ever show up in orderbys? 

We should sanity check (in tests):
- ORDER BY vector, labels
- ORDER BY labels, vector
- WHERE vector = $1 ORDER BY labels